// fp-ts-std/index.js
"use strict";
/**
 * All modules re-exported in a single primary entrypoint.
 *
 * @since 0.5.1
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.url = exports.urlSearchParams = exports.task = exports.string = exports.record = exports.option = exports.number = exports.json = exports.function = exports.either = exports.date = exports.boolean = exports.array = void 0;
const array = require("./Array");
exports.array = array;
const boolean = require("./Boolean");
exports.boolean = boolean;
const date = require("./Date");
exports.date = date;
const either = require("./Either");
exports.either = either;
const function_ = require("./Function");
exports.function = function_;
const json = require("./JSON");
exports.json = json;
const number = require("./Number");
exports.number = number;
const option = require("./Option");
exports.option = option;
const record = require("./Record");
exports.record = record;
const string = require("./String");
exports.string = string;
const task = require("./Task");
exports.task = task;
const urlSearchParams = require("./URLSearchParams");
exports.urlSearchParams = urlSearchParams;
const url = require("./URL");
exports.url = url;
