// name-wrapper/hardhat.config.js
const { utils } = require('ethers')
const fs = require('fs')
const { execSync } = require('child_process');

require('@nomiclabs/hardhat-waffle')
require('hardhat-gas-reporter')
require("@nomiclabs/hardhat-etherscan");

const commit = execSync('git rev-parse --short HEAD', ).toString().trim();
const privateKey1 = '0x2002fe06e0e71f997f6c0ec089f56682355c3ef54b0ecb72a60546c8ccbc7767'
module.exports = {
  solidity: {
    version: '0.8.4',
    settings: {
      optimizer: {
        enabled: true,
        runs: 200,
      },
    },
  },
  networks: {
    localhost: {
      url: 'http://localhost:8545',
    },
    rinkeby: {
      url: "https://rinkeby.infura.io/v3/2885a28cc0444998a0623d7b21fdf254",
      // gasPrice: 1000000000, // 5 gwei
      // gas:100000,
      accounts: [privateKey1]
    },
    goerli: {
      url: "https://goerli.infura.io/v3/2885a28cc0444998a0623d7b21fdf254",
      gasPrice: 500000000000,// 500 gwei
      gas:1000000,
      accounts: [privateKey1]
    }
  },
  etherscan: {
    // Your API key for Etherscan
    // Obtain one at https://etherscan.io/
    apiKey: "AN9EQIRIU7JUFQSKN25J2V37NMMB4DIMVI"
  },
  gasReporter: {
    excludeContracts: ['mocks', 'registry', 'ethregistrar'],
    outputFile: `gasreport-${commit}.txt`,
    noColors: true,
  }
}
